File: python/src/hhat_lang/core/config/quantum_device.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Literal

from hhat_lang.core.config.utils import read_file

QuantumDeviceKind = Literal["simulator", "device"]
TomlValue = str | int | float | bool | list[Any] | dict[str, Any] | None


@dataclass(frozen=True)
class QuantumDeviceSpecs:
    max_n_qubits: int
    qubit_topology: TomlValue = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> QuantumDeviceSpecs:
        if "max_n_qubits" not in data:
            raise ValueError("quantum device specs must define 'max_n_qubits'")
        return cls(
            max_n_qubits=int(data["max_n_qubits"]),
            qubit_topology=data.get("qubit_topology"),
        )

    def as_dict(self) -> dict[str, TomlValue]:
        return {
            "max_n_qubits": self.max_n_qubits,
            "qubit_topology": self.qubit_topology,
        }


@dataclass(frozen=True)
class QuantumLowLevelLanguage:
    name: str
    package_name: str
    package_version: str
    file_extension: str

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> QuantumLowLevelLanguage:
        return cls(
            name=_require_text(data, "name"),
            package_name=_require_text(data, "package_name"),
            package_version=_require_text(data, "package_version"),
            file_extension=_require_text(data, "file_extension"),
        )

    def as_dict(self) -> dict[str, str]:
        return {
            "name": self.name,
            "package_name": self.package_name,
            "package_version": self.package_version,
            "file_extension": self.file_extension,
        }


@dataclass(frozen=True)
class QuantumDevice:
    name: str
    is_active: bool
    vendor: str
    device_kind: QuantumDeviceKind
    device_platform: str
    computation_paradigm: tuple[str, ...]
    specs: QuantumDeviceSpecs
    qll: QuantumLowLevelLanguage
    extra: dict[str, TomlValue] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> QuantumDevice:
        kind = _require_text(data, "device_kind")
        if kind not in {"simulator", "device"}:
            raise ValueError("device_kind must be 'simulator' or 'device'")

        return cls(
            name=_require_text(data, "name"),
            is_active=bool(data.get("is_active", False)),
            vendor=_require_text(data, "vendor"),
            device_kind=kind,
            device_platform=_require_text(data, "device_platform"),
            computation_paradigm=tuple(str(value) for value in data.get("computation_paradigm", ())),
            specs=QuantumDeviceSpecs.from_dict(data.get("specs", {})),
            qll=QuantumLowLevelLanguage.from_dict(data.get("qll", {})),
            extra=dict(data.get("extra", {})),
        )

    def as_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "is_active": self.is_active,
            "vendor": self.vendor,
            "device_kind": self.device_kind,
            "device_platform": self.device_platform,
            "computation_paradigm": list(self.computation_paradigm),
            "specs": self.specs.as_dict(),
            "qll": self.qll.as_dict(),
            "extra": self.extra,
        }


@dataclass(frozen=True)
class QuantumDeviceConfig:
    title: str
    version: str
    devices: tuple[QuantumDevice, ...]

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> QuantumDeviceConfig:
        devices = tuple(QuantumDevice.from_dict(device) for device in data.get("devices", ()))
        config = cls(
            title=_require_text(data, "title"),
            version=_require_text(data, "version"),
            devices=devices,
        )
        config.validate()
        return config

    @property
    def active_devices(self) -> tuple[QuantumDevice, ...]:
        return tuple(device for device in self.devices if device.is_active)

    def as_dict(self) -> dict[str, Any]:
        return {
            "title": self.title,
            "version": self.version,
            "devices": [device.as_dict() for device in self.devices],
        }

    def validate(self) -> None:
        if not self.devices:
            raise ValueError("quantum device config must contain at least one device")
        if not self.active_devices:
            raise ValueError("quantum device config must contain at least one active device")


def load_quantum_config(file: Path | str = "quantum_config.toml") -> QuantumDeviceConfig:
    return QuantumDeviceConfig.from_dict(read_file(Path(file)))


def write_quantum_config(config: QuantumDeviceConfig, file: Path | str = "quantum_config.toml") -> None:
    config.validate()
    Path(file).write_text(_render_quantum_config(config))


def _render_quantum_config(config: QuantumDeviceConfig) -> str:
    lines = [
        f"title = {_toml_value(config.title)}",
        f"version = {_toml_value(config.version)}",
        "",
    ]

    for device in config.devices:
        lines.extend(
            [
                "[[devices]]",
                f"name = {_toml_value(device.name)}",
                f"is_active = {_toml_value(device.is_active)}",
                f"vendor = {_toml_value(device.vendor)}",
                f"device_kind = {_toml_value(device.device_kind)}",
                f"device_platform = {_toml_value(device.device_platform)}",
                f"computation_paradigm = {_toml_value(list(device.computation_paradigm))}",
                "[devices.specs]",
                f"max_n_qubits = {_toml_value(device.specs.max_n_qubits)}",
            ]
        )
        if device.specs.qubit_topology is not None:
            lines.append(f"qubit_topology = {_toml_value(device.specs.qubit_topology)}")

        lines.extend(
            [
                "[devices.qll]",
                f"name = {_toml_value(device.qll.name)}",
                f"package_name = {_toml_value(device.qll.package_name)}",
                f"package_version = {_toml_value(device.qll.package_version)}",
                f"file_extension = {_toml_value(device.qll.file_extension)}",
                "[devices.extra]",
            ]
        )
        for key, value in sorted(device.extra.items()):
            lines.append(f"{key} = {_toml_value(value)}")
        lines.append("")

    return "\n".join(lines)


def _require_text(data: dict[str, Any], key: str) -> str:
    value = data.get(key)
    if not isinstance(value, str) or not value:
        raise ValueError(f"quantum device config must define non-empty '{key}'")
    return value


def _toml_value(value: TomlValue) -> str:
    match value:
        case bool():
            return str(value).lower()
        case int() | float():
            return str(value)
        case str():
            return '"' + value.replace("\\", "\\\\").replace('"', '\\"') + '"'
        case list() | tuple():
            return "[" + ", ".join(_toml_value(item) for item in value) + "]"
        case dict():
            items = ", ".join(f"{key} = {_toml_value(item)}" for key, item in sorted(value.items()))
            return "{ " + items + " }"
        case None:
            raise ValueError("None is not a valid TOML value")
        case _:
            raise TypeError(f"unsupported TOML value type: {type(value).__name__}")

File: rust/hhat_lang/src/config/quantum_device.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::error::Error;
use std::fmt;
use std::fs;
use std::path::Path;

pub const DEFAULT_QUANTUM_CONFIG_FILE: &str = "quantum_config.toml";

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct QuantumDeviceConfig {
    pub title: String,
    pub version: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub available_devices: Vec<String>,
    #[serde(default)]
    pub devices: Vec<QuantumDevice>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct QuantumDevice {
    pub name: String,
    pub is_active: bool,
    pub vendor: String,
    pub device_kind: DeviceKind,
    pub device_platform: String,
    #[serde(default)]
    pub computation_paradigm: Vec<String>,
    pub specs: QuantumDeviceSpecs,
    pub qll: QuantumLowLevelLanguage,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub extra: BTreeMap<String, toml::Value>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DeviceKind {
    Simulator,
    Device,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct QuantumDeviceSpecs {
    pub max_n_qubits: u32,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub qubit_topology: BTreeMap<String, toml::Value>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct QuantumLowLevelLanguage {
    pub name: String,
    pub package_name: String,
    pub package_version: String,
    pub file_extension: String,
}

#[derive(Debug)]
pub enum QuantumConfigError {
    Io(std::io::Error),
    Decode(toml::de::Error),
    Encode(toml::ser::Error),
    Validation(String),
}

impl fmt::Display for QuantumConfigError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(error) => write!(f, "quantum device config I/O error: {error}"),
            Self::Decode(error) => write!(f, "invalid quantum device config TOML: {error}"),
            Self::Encode(error) => write!(f, "could not encode quantum device config: {error}"),
            Self::Validation(message) => write!(f, "invalid quantum device config: {message}"),
        }
    }
}

impl Error for QuantumConfigError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Io(error) => Some(error),
            Self::Decode(error) => Some(error),
            Self::Encode(error) => Some(error),
            Self::Validation(_) => None,
        }
    }
}

impl From<std::io::Error> for QuantumConfigError {
    fn from(error: std::io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<toml::de::Error> for QuantumConfigError {
    fn from(error: toml::de::Error) -> Self {
        Self::Decode(error)
    }
}

impl From<toml::ser::Error> for QuantumConfigError {
    fn from(error: toml::ser::Error) -> Self {
        Self::Encode(error)
    }
}

impl QuantumDeviceConfig {
    pub fn new(title: impl Into<String>, version: impl Into<String>) -> Self {
        Self {
            title: title.into(),
            version: version.into(),
            available_devices: Vec::new(),
            devices: Vec::new(),
        }
    }

    pub fn read_from_file(path: impl AsRef<Path>) -> Result<Self, QuantumConfigError> {
        let mut config: Self = toml::from_str(&fs::read_to_string(path)?)?;
        config.refresh_available_devices();
        config.validate()?;
        Ok(config)
    }

    pub fn write_to_file(&self, path: impl AsRef<Path>) -> Result<(), QuantumConfigError> {
        let mut config = self.clone();
        config.refresh_available_devices();
        config.validate()?;

        if let Some(parent) = path
            .as_ref()
            .parent()
            .filter(|path| !path.as_os_str().is_empty())
        {
            fs::create_dir_all(parent)?;
        }

        fs::write(path, toml::to_string_pretty(&config)?)?;
        Ok(())
    }

    pub fn validate(&self) -> Result<(), QuantumConfigError> {
        if self.title.trim().is_empty() {
            return Err(QuantumConfigError::Validation(
                "title must not be empty".to_string(),
            ));
        }
        if self.version.trim().is_empty() {
            return Err(QuantumConfigError::Validation(
                "version must not be empty".to_string(),
            ));
        }
        if self.devices.is_empty() {
            return Err(QuantumConfigError::Validation(
                "at least one quantum device must be configured".to_string(),
            ));
        }
        if self.active_devices().is_empty() {
            return Err(QuantumConfigError::Validation(
                "at least one quantum device must be active".to_string(),
            ));
        }

        for device in &self.devices {
            device.validate()?;
        }

        Ok(())
    }

    pub fn add_device(&mut self, device: QuantumDevice) {
        self.devices.push(device);
        self.refresh_available_devices();
    }

    pub fn upsert_device(&mut self, device: QuantumDevice) {
        if let Some(existing) = self
            .devices
            .iter_mut()
            .find(|existing| existing.name == device.name)
        {
            *existing = device;
        } else {
            self.devices.push(device);
        }
        self.refresh_available_devices();
    }

    pub fn active_devices(&self) -> Vec<&QuantumDevice> {
        self.devices
            .iter()
            .filter(|device| device.is_active)
            .collect()
    }

    pub fn active_device_names(&self) -> Vec<String> {
        self.active_devices()
            .into_iter()
            .map(|device| device.name.clone())
            .collect()
    }

    pub fn find_device(&self, name: &str) -> Option<&QuantumDevice> {
        self.devices.iter().find(|device| device.name == name)
    }

    pub fn find_active_device(&self, name: &str) -> Option<&QuantumDevice> {
        self.find_device(name).filter(|device| device.is_active)
    }

    pub fn refresh_available_devices(&mut self) {
        self.available_devices = self.active_device_names();
    }
}

impl Default for QuantumDeviceConfig {
    fn default() -> Self {
        Self::new("H-hat Compiler Quantum Device Configuration", "0.1.0")
    }
}

impl QuantumDevice {
    pub fn validate(&self) -> Result<(), QuantumConfigError> {
        if self.name.trim().is_empty() {
            return Err(QuantumConfigError::Validation(
                "device name must not be empty".to_string(),
            ));
        }
        if self.vendor.trim().is_empty() {
            return Err(QuantumConfigError::Validation(format!(
                "device '{}' must include a vendor",
                self.name
            )));
        }
        if self.device_platform.trim().is_empty() {
            return Err(QuantumConfigError::Validation(format!(
                "device '{}' must include a platform",
                self.name
            )));
        }
        if self.computation_paradigm.is_empty() {
            return Err(QuantumConfigError::Validation(format!(
                "device '{}' must include at least one computation paradigm",
                self.name
            )));
        }
        if self.specs.max_n_qubits == 0 {
            return Err(QuantumConfigError::Validation(format!(
                "device '{}' must expose at least one qubit",
                self.name
            )));
        }
        self.qll.validate(&self.name)?;
        Ok(())
    }
}

impl QuantumLowLevelLanguage {
    fn validate(&self, device_name: &str) -> Result<(), QuantumConfigError> {
        if self.name.trim().is_empty()
            || self.package_name.trim().is_empty()
            || self.package_version.trim().is_empty()
            || self.file_extension.trim().is_empty()
        {
            return Err(QuantumConfigError::Validation(format!(
                "device '{device_name}' must include complete QLL information"
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{SystemTime, UNIX_EPOCH};

    fn topology(entries: &[(&str, toml::Value)]) -> BTreeMap<String, toml::Value> {
        entries
            .iter()
            .map(|(key, value)| ((*key).to_string(), value.clone()))
            .collect()
    }

    fn device(
        name: &str,
        is_active: bool,
        vendor: &str,
        kind: DeviceKind,
        platform: &str,
        paradigms: &[&str],
        max_n_qubits: u32,
        qll_name: &str,
        package_name: &str,
        package_version: &str,
        file_extension: &str,
    ) -> QuantumDevice {
        QuantumDevice {
            name: name.to_string(),
            is_active,
            vendor: vendor.to_string(),
            device_kind: kind,
            device_platform: platform.to_string(),
            computation_paradigm: paradigms
                .iter()
                .map(|paradigm| (*paradigm).to_string())
                .collect(),
            specs: QuantumDeviceSpecs {
                max_n_qubits,
                qubit_topology: topology(&[("kind", toml::Value::String("all-to-all".into()))]),
            },
            qll: QuantumLowLevelLanguage {
                name: qll_name.to_string(),
                package_name: package_name.to_string(),
                package_version: package_version.to_string(),
                file_extension: file_extension.to_string(),
            },
            extra: BTreeMap::new(),
        }
    }

    fn dummy_config() -> QuantumDeviceConfig {
        let mut config = QuantumDeviceConfig::default();
        config.add_device(device(
            "AerSimulator",
            true,
            "IBM",
            DeviceKind::Simulator,
            "superconducting",
            &["gate"],
            32,
            "openqasm",
            "qiskit",
            "1.0.0",
            "qasm",
        ));
        config.add_device(device(
            "default.qubit",
            true,
            "Xanadu",
            DeviceKind::Simulator,
            "statevector",
            &["gate"],
            16,
            "pennylane",
            "pennylane",
            "0.40.0",
            "py",
        ));
        config.add_device(device(
            "H1-1",
            false,
            "Quantinuum",
            DeviceKind::Device,
            "trapped ions",
            &["gate"],
            20,
            "openqasm",
            "pytket-quantinuum",
            "0.37.0",
            "qasm",
        ));
        config.add_device(device(
            "Aquila",
            false,
            "QuEra",
            DeviceKind::Device,
            "neutral atoms",
            &["analog"],
            256,
            "bloqade",
            "bloqade",
            "0.18.0",
            "json",
        ));
        config
    }

    fn temp_config_path() -> std::path::PathBuf {
        let stamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("system clock should be after unix epoch")
            .as_nanos();
        std::env::temp_dir()
            .join(format!("hhat_quantum_config_test_{stamp}"))
            .join(DEFAULT_QUANTUM_CONFIG_FILE)
    }

    #[test]
    fn reads_and_writes_dummy_quantum_devices() {
        let path = temp_config_path();
        let config = dummy_config();

        config
            .write_to_file(&path)
            .expect("dummy config should be writable");
        let loaded =
            QuantumDeviceConfig::read_from_file(&path).expect("dummy config should be readable");

        assert_eq!(loaded.title, "H-hat Compiler Quantum Device Configuration");
        assert_eq!(loaded.version, "0.1.0");
        assert_eq!(loaded.devices.len(), 4);
        assert_eq!(
            loaded.available_devices,
            vec!["AerSimulator".to_string(), "default.qubit".to_string()]
        );
        assert_eq!(
            loaded
                .find_active_device("AerSimulator")
                .expect("AerSimulator should be active")
                .qll
                .package_name,
            "qiskit"
        );
        assert!(loaded.find_active_device("H1-1").is_none());

        let _ = fs::remove_file(&path);
        let _ = path.parent().map(fs::remove_dir);
    }

    #[test]
    fn derives_available_devices_when_missing_from_toml() {
        let toml = r#"
title = "H-hat Compiler Quantum Device Configuration"
version = "0.1.0"

[[devices]]
name = "AerSimulator"
is_active = true
vendor = "IBM"
device_kind = "simulator"
device_platform = "superconducting"
computation_paradigm = ["gate"]

[devices.specs]
max_n_qubits = 32
qubit_topology = { kind = "all-to-all" }

[devices.qll]
name = "openqasm"
package_name = "qiskit"
package_version = "1.0.0"
file_extension = "qasm"
"#;

        let mut config: QuantumDeviceConfig =
            toml::from_str(toml).expect("manual quantum config should parse");
        config.refresh_available_devices();
        config.validate().expect("manual quantum config is valid");

        assert_eq!(config.available_devices, vec!["AerSimulator".to_string()]);
    }

    #[test]
    fn rejects_configs_without_active_devices() {
        let mut config = QuantumDeviceConfig::default();
        config.add_device(device(
            "H1-1",
            false,
            "Quantinuum",
            DeviceKind::Device,
            "trapped ions",
            &["gate"],
            20,
            "openqasm",
            "pytket-quantinuum",
            "0.37.0",
            "qasm",
        ));

        let error = config
            .validate()
            .expect_err("config without active devices should fail");

        assert!(error
            .to_string()
            .contains("at least one quantum device must be active"));
    }

    #[test]
    fn upsert_device_refreshes_available_devices() {
        let mut config = dummy_config();
        config.upsert_device(device(
            "Aquila",
            true,
            "QuEra",
            DeviceKind::Device,
            "neutral atoms",
            &["analog"],
            256,
            "bloqade",
            "bloqade",
            "0.18.0",
            "json",
        ));

        assert_eq!(config.devices.len(), 4);
        assert!(config.available_devices.contains(&"Aquila".to_string()));
        assert_eq!(
            config
                .find_active_device("Aquila")
                .expect("Aquila should have been activated")
                .specs
                .max_n_qubits,
            256
        );
    }
}
